Sources/Network/JSON.swift

```diff
@@ -262,6 +262,62 @@ public enum JSON {
+    /// An exception is thrown if the formatter is unable to convert the provided value into Date.
+    /// Additionally, an optional `parseAPIError` closure can be supplied so that if a parsing step fails
+    /// an attempt is made to extract a domain specific error and throw it.
+    ///
+    /// - Parameters:
+    ///   - key: The JSON attribute key to parse.
+    ///   - json: The JSON dictionary.
+    ///   - formatter: The formatter to convert the parsed type into Date.
+    ///   - parseAPIError: The API error parsing closure.
+    /// - Returns: The Date obtained from the formatter.
+    /// - Throws: An error of type `JSON.Error`, or a domain specific `Swift.Error` produced by `parseAPIError`.
+    public static func parseDateAttribute<T>(key: JSON.AttributeKey,
+                                             json: JSON.Dictionary,
+                                             formatter: (T) -> Date?,
+                                             parseAPIError: ParseAPIErrorClosure? = nil) throws -> Date {
+
+        guard let parsedDate = try parseOptionalDateAttribute(key: key,
+                                                          json: json,
+                                                          formatter: formatter,
+                                                          parseAPIError: parseAPIError) else {
+            throw Error.unexpectedAttributeValue(key, json: json)
+        }
+
+        return parsedDate
+    }
+
+
+    /// Parse an attribute of type `T` with the given key on the given JSON dictionary, trying to convert it to a date,
+    /// using the provided formatter which receives a `T` and returns a Date or nil if it fails.
+    /// Additionally, an optional `parseAPIError` closure can be supplied so that if a parsing step fails
+    /// an attempt is made to extract a domain specific error and throw it.
+    ///
+    /// - Parameters:
+    ///   - key: The JSON attribute key to parse.
+    ///   - json: The JSON dictionary.
+    ///   - formatter: The formatter to convert the parsed type into Date.
+    ///   - parseAPIError: The API error parsing closure.
+    /// - Returns: The Date obtained from the formatter or nil.
+    /// - Throws: An error of type `JSON.Error`, or a domain specific `Swift.Error` produced by `parseAPIError`.
+    public static func parseOptionalDateAttribute<T>(key: JSON.AttributeKey,
+                                                     json: JSON.Dictionary,
+                                                     formatter: (T) -> Date?,
+                                                     parseAPIError: ParseAPIErrorClosure? = nil) throws -> Date? {
+
+        guard let rawValue = json[key] else {
+            throw parseAPIError?(json) ?? Error.missingAttribute(key, json: json)
+        }
+
+        let jsonValue: T = try parseValue(rawValue: rawValue, key: key, json: json, parseAPIError: parseAPIError)
+
+        return formatter(jsonValue)
```

Comment:
What if the formatter fails? Shouldn't we throw an exception? 

e.g.: `throw Error.unexpectedAttributeValue(key, json: json)`

Reply:
But this is the case where it can be Optional, right? so the formatter may not be able to convert it, fair enough, but it returns nil. 
it doesn't seems wrong to me

Reply:
From my understanding, the `parseOptionalAttribute` should only return `nil` if the JSON doesn't contain a value for that key. If it does contain the value, it should essentially behave like a normal `parseAttribute`.

So in this case, I think we should preserve the behavior from the other one in regards to invoking `formatter()` 😜

Reply:
agree ✅